File: 백준/Bronze/2798. 블랙잭/블랙잭.js
const input = require("fs").readFileSync("/dev/stdin").toString().split("\n");
const [n, maxData] = input[0].trim().split(" ").map((e) => parseInt(e));
const arr = input[1].split(" ").map((e) => parseInt(e));
const getCombinations = (array, selectNumber) => {
    const results = [];
    if(selectNumber === 1){
        return array.map((element) => [element]);
    }
    array.forEach((fixed, index, origin) => {
        const rest = origin.slice(index+1);
        const combinations = getCombinations(rest, selectNumber - 1);
        const attached = combinations.map((combination) => [fixed, ...combination]);
        results.push(...attached);
    });
    return results;
}

let add = arr => arr.reduce((a, b) => a + b, 0);

let result = 0;

getCombinations(arr, 3).forEach((elem) => {
    let temp = add(elem);
    if (temp <= maxData && temp > result) {
        result = temp;
    }
})

console.log(result);

File: 백준/Bronze/2798. 블랙잭/블랙잭.py
import sys, itertools
N, M = map(int,sys.stdin.readline().split(" "))
datas = list(map(int,sys.stdin.readline().split(" ")))
min_data = 100000
result = []
# M을 넘지 않으면서 M에 최대한 가까운 카드 3장
for _ in itertools.combinations(datas,3):
    tmp = sum(_)
    if tmp <= M and M-tmp <= min_data:
        min_data = M-tmp
        result = _
print(sum(result))
